src/h5web/visualizations/shared/utils.ts

```diff
@@ -90,13 +90,59 @@ function useAxisScale(config: AxisConfig, rangeSize: number): AxisScale {
 export function useAbscissaScale(): AxisScale {
   const { size } = useThree();
   const { abscissaConfig } = useContext(AxisSystemContext);
-  return useAxisScale(abscissaConfig, size.width);
+  return getAxisScale(abscissaConfig, size.width);
 }
 
 export function useOrdinateScale(): AxisScale {
   const { size } = useThree();
   const { ordinateConfig } = useContext(AxisSystemContext);
-  return useAxisScale(ordinateConfig, size.height);
+  return getAxisScale(ordinateConfig, size.height);
+}
+
+/**
+ * We can't rely on the axis scale's `ticks` method to get integer tick values because
+ * `d3.tickStep` sometimes returns incoherent step values - e.g. `d3.tickStep(0, 2, 3)`
+ * returns 0.5 intead of 1.
+ *
+ * So we implement our own, simplified version of `d3.ticks` that always outputs integer values.
+ */
+function getIntegerTicks(domain: Domain, count: number): number[] {
+  const [min, max] = domain;
+  const intMin = Math.ceil(min);
+  const intMax = Math.floor(max);
+
+  const domainLength = intMax - intMin + 1;
+  const optimalCount = Math.min(domainLength, count);
+
+  if (optimalCount === 0) {
+    return [];
+  }
+
+  // Make sure we always use a step that is greater than or equal to 1
+  const step = Math.max(tickStep(intMin, intMax, optimalCount), 1);
+
+  const start = Math.ceil(min / step);
+  const stop = Math.floor(max / step);
+  const len = Math.ceil(stop - start + 1);
+
+  const ticks = new Array(len);
+  for (let i = 0; i < len; i++) {
+    ticks[i] = (start + i) * step;
+  }
+
+  return ticks;
+}
+
+export function getTicksProp(
+  config: AxisConfig,
+  visibleDomain: Domain,
+  availableSize: number
+): { tickValues: number[] } | { numTicks: number } {
+  const numTicks = adaptedNumTicks(availableSize);
+
+  return isIndexAxisConfig(config)
+    ? { tickValues: getIntegerTicks(visibleDomain, numTicks) }
+    : { numTicks };
```

Comment:
For the last tick values, you could put a boolean `hideLastTickValue` in `config` and check here if you need the slice the result from `getIntegerTicks(visibleDomain, numTicks)` :shrug:

Reply:
As discussed, see #144